### PR TITLE
ED-1542 Skeleton of Behave project

### PR DESCRIPTION
**Pull request by kowalcj0**
This contains a working `Behave` project with an example test scenario that verifies whether Supplier can create a SSO account for selected company.
It mimics every step made by real User when clicking through FAB and SSO pages.

PS. This branch should be actually linked with https://uktrade.atlassian.net/browse/ED-1659 not ED-1542

### CODE REVIEW

File: tests/functional/.gitignore [hunk 1]
```diff
@@ -0,0 +1,80 @@
+# Created by https://www.gitignore.io
```
Comment (konradko) [inline tests/functional/.gitignore]: why separate?

File: tests/functional/features/ScenarioData.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
+# -*- coding: utf-8 -*-
```
Comment (konradko) [inline tests/functional/features/ScenarioData.py]: camel case :(

**Comment by konradko** [comment]
It looks like you're setting up a separate repository here - maybe would be better to just have `directory-tests-functional`?

File: tests/functional/features/ScenarioData.py (hunk 2 above)
Comment (kowalcj0) [inline tests/functional/features/ScenarioData.py]: it was initially scoped as a class, will rename it

File: tests/functional/.gitignore (hunk 1 above)
Comment (kowalcj0) [inline tests/functional/.gitignore]: copy-pasta

**Comment by kowalcj0** [comment]
@konradko I thought about have a dedicated repo for this project, but then I've noticed that I can make use of endpoint definitions stored in `./tests/__init__.py`. 
Apart from that I can't think of any benefit of having a separate repo.

File: tests/functional/README.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
+Functional Test for FAB, FAS & SSO
```
Comment (konradko) [inline tests/functional/README.md]: this should be a part of the main readme, no reason to hide it away :)

File: tests/functional/features/context_utils.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+"""Helpers added to the behave's `context` object."""
+import logging
+from collections import namedtuple
+from types import MethodType
+
+from requests import Session
+
+
+ScenarioData = namedtuple('ScenarioData', ['actors', 'unregistered_companies'])
+Actor = namedtuple('Actor', ['alias', 'email', 'password', 'session',
```
Comment (konradko) [inline tests/functional/features/context_utils.py]: preferred line breaking method (line up under the first character of the line that starts the multi-line construct): 
```
namedtuple(
    'Actor',
    [
        'alias', 'email', 'password', 'session',
        'csrfmiddlewaretoken', 'email_confirmation_link'
    ]
)
```

File: tests/functional/README.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+Both methods described below require all env variables to be set.
+In case of the IDE runner, you can set all required environment variables in
+the `run configuration`.
+
+From CLI:
+```shell
+behave tests/functional/features/
+```
+
+From IDE (ideally from PyCharm Pro):
```
Comment (konradko) [inline tests/functional/README.md]: If the instructions are only tested under Pycharm it's fine to just state `From Pycharm`

File: tests/functional/features/context_utils.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
+# -*- coding: utf-8 -*-
+"""Helpers added to the behave's `context` object."""
+import logging
+from collections import namedtuple
+from types import MethodType
+
+from requests import Session
+
+
+ScenarioData = namedtuple('ScenarioData', ['actors', 'unregistered_companies'])
```
Comment (konradko) [inline tests/functional/features/context_utils.py]: `lower_case_with_underscores` for variables please - we follow PEP8 here :P

File: tests/functional/README.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+# create virtualenv & install dependencies
+```bash
+mkvirtualenv .env -p python3.5
+pip install -r requirements.txt
+```
+
+# updating the requirements.txt
+
+This project uses [pip-compile](https://pypi.python.org/pypi/pip-tools/) tool to generate `requirements.txt` with
```
Comment (konradko) [inline tests/functional/README.md]: this project does not use `pip-compile` or should use it consistently among all `requirements.txt` files.

Please either remove pip-compile here or make it work for all reqs in this repo + create a ticket to change the rest of `directory` repositories to use it. `pip-sync` is nice, we should use it.

File: tests/functional/features/context_utils.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    :return: an Actor named tuple
+    :rtype actor: tests.functional.features.ScenarioData.Actor
+    """
+    res = None
+    for actor in self.scenario_data.actors:
+        if actor.alias == alias:
+            res = actor
+            logging.debug("Found actor: '%s' in Scenario Data", alias)
+    assert res is not None, ("Couldn't find actor '{}' in Scenario Data"
+                             .format(alias))
```
Comment (konradko) [inline tests/functional/features/context_utils.py]: please use string formatting consistently - either `%s` or `format`. The latter is preferred

File: tests/functional/README.md (hunk 5 above)
Comment (kowalcj0) [inline tests/functional/README.md]: @konradko  Community version of `PyCharm` has some problems with parsing the `feature` files correctly, and thus I mentioned about `PyCharm Pro` which doesn't have such issues

File: tests/functional/features/context_utils.py [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+    return res
+
+
+def set_actor_csrfmiddlewaretoken(self, alias, token):
+    for idx, actor in enumerate(self.scenario_data.actors):
+        if actor.alias == alias:
+            self.scenario_data.actors[idx] = \
```
Comment (konradko) [inline tests/functional/features/context_utils.py]: we actively discourage using `\` for line breaking - it's ugly :)

File: tests/functional/README.md (hunk 7 above)
Comment (kowalcj0) [inline tests/functional/README.md]: @konradko you mean this [pip-sync](https://github.com/hugopeixoto/pip-sync)??? I don't how this is better than `pip-compile`.

File: tests/functional/features/context_utils.py (hunk 8 above)
Comment (kowalcj0) [inline tests/functional/features/context_utils.py]: yeah, you're right. @konradko

File: tests/functional/features/context_utils.py [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+        if actor.alias == alias:
+            res = actor
+            logging.debug("Found actor: '%s' in Scenario Data", alias)
+    assert res is not None, ("Couldn't find actor '{}' in Scenario Data"
+                             .format(alias))
+    return res
+
+
+def set_actor_csrfmiddlewaretoken(self, alias, token):
+    for idx, actor in enumerate(self.scenario_data.actors):
```
Comment (konradko) [inline tests/functional/features/context_utils.py]: Looks like you often iterate over a list to find matching value - using a map here would be more efficient and simple:
```
actors_by_alias = {alias: actor}
actors_by_alias[alias]._replace(csrfmiddlewaretoken=token)
```

File: tests/functional/features/settings.py [hunk 11]
```diff
@@ -0,0 +1,13 @@
+import os
+
+
+DIRECTORY_API_URL = os.environ["DIRECTORY_API_URL"]
+DIRECTORY_SSO_URL = os.environ['DIRECTORY_SSO_URL']
+DIRECTORY_UI_BUYER_URL = os.environ["DIRECTORY_UI_BUYER_URL"]
+DIRECTORY_UI_SUPPLIER_URL = os.environ["DIRECTORY_UI_SUPPLIER_URL"]
+DIRECTORY_PROFILE_URL = os.environ["DIRECTORY_PROFILE_URL"]
+
+S3_ACCESS_KEY_ID = os.environ["S3_ACCESS_KEY_ID"]
```
Comment (konradko) [inline tests/functional/features/settings.py]: comment why this is needed would be useful

File: tests/functional/features/context_utils.py (hunk 10 above)
Comment (kowalcj0) [inline tests/functional/features/context_utils.py]: @konradko maybe filter would be better?
```
def find_actor(self, alias):
    actors = list(filter((lambda x: x.alias == alias, self.scenario_data.actors)))
    assert actors
    return actors[0]
```

Comment (kowalcj0) [inline tests/functional/features/context_utils.py]: @konradko unfortunately `filter()` doesn't work well, as it looses reference to the original object stored in the `context.scenario_data.actors`, 

```python
# import Actor named tuple
from tests.functional.features.context_utils import Actor

# create a list of few Actors to simulate state of `context.scenario_data.actors`
actors = [Actor(alias="a",email=None,password=None,session=None,csrfmiddlewaretoken=None,email_confirmation_link=None), Actor(alias="b",email=None,password=None,session=None,csrfmiddlewaretoken=None,email_confirmation_link=None), Actor(alias="c",email=None,password=None,session=None,csrfmiddlewaretoken=None,email_confirmation_link=None)]

# check the state of actors
actors
[Actor(alias='a', email=None, password=None, session=None, csrfmiddlewaretoken=None, email_confirmation_link=None),
 Actor(alias='b', email=None, password=None, session=None, csrfmiddlewaretoken=None, email_confirmation_link=None),
 Actor(alias='c', email=None, password=None, session=None, csrfmiddlewaretoken=None, email_confirmation_link=None)]

#filter out sought actor
found_actor = list(filter((lambda x: x.alias=='b'), actors))[0]

# see it's details
found_actor
Actor(alias='b', email=None, password=None, session=None, csrfmiddlewaretoken=None, email_confirmation_link=None)

# use `_replace()` to change value of a named tuple property 
found_actor = found_actor._replace(email_confirmation_link="some value")

# check the found actor
found_actor
Actor(alias='b', email=None, password=None, session=None, csrfmiddlewaretoken=None, email_confirmation_link='some value')

# check the Actors
# Actor stored in context.scenario_data.actors remained unchanged: 
actors
[Actor(alias='a', email=None, password=None, session=None, csrfmiddlewaretoken=None, email_confirmation_link=None),
 Actor(alias='b', email=None, password=None, session=None, csrfmiddlewaretoken=None, email_confirmation_link=None),
 Actor(alias='c', email=None, password=None, session=None, csrfmiddlewaretoken=None, email_confirmation_link=None)]
```

File: tests/functional/features/context_utils.py (hunk 8 above)
Comment (kowalcj0) [inline tests/functional/features/context_utils.py]: @konradko as per using `.format()` over `%`, I also prefer `.format()` but I'm slowly leaning towards `%`. Please have a look at these two pages:
https://stackoverflow.com/a/34634301
https://docs.python.org/3/howto/logging.html#optimization

Moreover `pylint` complains about using `.format()` in in logging functions. Here's an example:
`Use % formatting in logging functions and pass the % parameters as arguments (logging-format-interpolation)`

**Comment by kowalcj0** [comment]
I used: `isort -rc -w 80 -l 80 -m 3` to sort and format imports.

File: tests/functional/README.md (hunk 7 above)
Comment (kowalcj0) [inline tests/functional/README.md]: ps. I'd prefer to leave as is as there's a chance that we'll me moving this to a dedicated repo.

File: tests/functional/README.md (hunk 3 above)
Comment (kowalcj0) [inline tests/functional/README.md]: I'd prefer to leave as is as there's a chance that we'll me moving this to a dedicated repo.

File: tests/functional/features/context_utils.py (hunk 8 above)
Comment (konradko) [inline tests/functional/features/context_utils.py]: we use flake8 :P Point taken though, solid research!

File: tests/functional/features/context_utils.py (hunk 9 above)
Comment (kowalcj0) [inline tests/functional/features/context_utils.py]: @konradko Fixed that in #39 https://github.com/uktrade/directory-tests/pull/39/commits/180c101adfabdfeb2d40d3c6e42ce8b1e2e93ba2

File: tests/functional/README.md (hunk 3 above)
Comment (konradko) [inline tests/functional/README.md]: Then let's separate it at that point - it should be immediately clear what the repo contains / how to use it when you go to its main page.

File: tests/functional/features/context_utils.py (hunk 10 above)
Comment (konradko) [inline tests/functional/features/context_utils.py]: Sorry, by `map` I meant Python's `dictionary`. To change the value you would

```
actors_by_alias = {alias: actor}
actors_by_alias[alias] = actors_by_alias[alias] ._replace(csrfmiddlewaretoken=token)
```

File: tests/functional/README.md (hunk 3 above)
Comment (kowalcj0) [inline tests/functional/README.md]: @konradko I removed this README and move most important parts of it to the main README
I also got rid of the `pip-compile` stuff.

Comment (konradko) [inline tests/functional/README.md]: 👍

File: tests/functional/features/utils.py [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+    :type  trim_response_content: bool
+    :return: a response object
+    :rtype: requests.Response
+    """
+    assert url is not None, "Please provide the URL"
+
+    req = session or requests
+
+    if method == Method.DELETE:
+        res = req.delete(url=url, params=params, headers=headers,
```
Comment (konradko) [inline tests/functional/features/utils.py]: bit more DRY:
request_kwargs = dict(url=url, params=params, headers=headers, cookies=cookies, allow_redirects=allow_redirects)

if method == Method.DELETE:
    res = req.delete(**request_kwargs)
```

File: tests/functional/features/utils.py [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+        logging.debug("REQ Data: %s", res.request.body)
+    if files:
+        logging.debug("REQ Files: %s", res.request.body)
+    logging.debug("RSP Status: %s %s", res.status_code, res.reason)
+    logging.debug("RSP URL: %s", res.url)
+    logging.debug("RSP Headers: %s", res.headers)
+    logging.debug("RSP Cookies: %s", res.cookies)
+    if res.content:
+        if trim_response_content:
+            if len(res.content) > 150:
```
Comment (konradko) [inline tests/functional/features/utils.py]: magic numbers should generally be settings

File: tests/functional/features/utils.py [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+
+    :param bucket: S3 bucket connection
+    :param actor: Actor named tuple
+    :param subject: expected subject of sought message
+    :return: a plain text message payload
+    """
+    res = None
+    found = False
+    for key in bucket.list():
+        if key.key != "AMAZON_SES_SETUP_NOTIFICATION":
```
Comment (konradko) [inline tests/functional/features/utils.py]: this message can just be deleted

File: tests/functional/features/utils.py [hunk 15]
```diff
@@ -0,0 +1,305 @@
...
+            logging.debug("Processing email file: %s", key.key)
+            try:
+                msg_contents = key.get_contents_as_string().decode("utf-8")
+                msg = email.message_from_string(msg_contents)
+                if msg['To'] == actor.email:
```
Comment (konradko) [inline tests/functional/features/utils.py]: are you using UUID for `test+<unique string>@`?

File: tests/functional/features/utils.py (hunk 12 above)
Comment (kowalcj0) [inline tests/functional/features/utils.py]: :+1:

File: tests/functional/features/utils.py (hunk 13 above)
Comment (kowalcj0) [inline tests/functional/features/utils.py]: @konradko Will make it a variable. This value is used only locally to specify how much of the response content should be logged.

File: tests/functional/features/utils.py (hunk 14 above)
Comment (kowalcj0) [inline tests/functional/features/utils.py]: @konradko I didn't know it's not important.

Comment (kowalcj0) [inline tests/functional/features/utils.py]: done :)

File: tests/functional/features/utils.py (hunk 15 above)
Comment (kowalcj0) [inline tests/functional/features/utils.py]: @konradko Nope, I was using `random.randint(100000, 999999)` but can replace it with `str(uuid.uuid4())`

Comment (kowalcj0) [inline tests/functional/features/utils.py]: Changed in https://github.com/uktrade/directory-tests/pull/44 -> https://github.com/uktrade/directory-tests/pull/44/commits/0161ab7f7f18a8949e43e6e1483af3ecef2ac81d

File: tests/functional/features/utils.py (hunk 14 above)
Comment (kowalcj0) [inline tests/functional/features/utils.py]: Done in https://github.com/uktrade/directory-tests/pull/44 -> https://github.com/uktrade/directory-tests/pull/44/commits/68676541a0aed8379bbd248cf067cddec8726823

File: tests/functional/features/utils.py (hunk 13 above)
Comment (kowalcj0) [inline tests/functional/features/utils.py]: Done in https://github.com/uktrade/directory-tests/pull/44 -> https://github.com/uktrade/directory-tests/pull/44/commits/5617779ac6d952775f2a1bd59f316586f7d69aaf

File: tests/functional/features/utils.py (hunk 12 above)
Comment (kowalcj0) [inline tests/functional/features/utils.py]: Done in https://github.com/uktrade/directory-tests/pull/44 -> https://github.com/uktrade/directory-tests/pull/44/commits/ad51e4ec938fbe9009adeb9b575a0b189e87c294

**Comment by kowalcj0** [comment]
@konradko In order to avoid all merging suggested changes up the streams I've applied them in https://github.com/uktrade/directory-tests/pull/44 
Here are the commits:
* https://github.com/uktrade/directory-tests/pull/44/commits/0161ab7f7f18a8949e43e6e1483af3ecef2ac81d
* https://github.com/uktrade/directory-tests/pull/44/commits/ad51e4ec938fbe9009adeb9b575a0b189e87c294
* https://github.com/uktrade/directory-tests/pull/44/commits/5617779ac6d952775f2a1bd59f316586f7d69aaf
* https://github.com/uktrade/directory-tests/pull/44/commits/68676541a0aed8379bbd248cf067cddec8726823

File: tests/functional/features/utils.py (hunk 15 above)
Comment (konradko) [inline tests/functional/features/utils.py]: 👍

**Pull request merged**